### PR TITLE
Redis Result Storage Engine

### PR DESCRIPTION
**Pull request by moizarafat**
## Description
Adds an implementation of Redis Result Storage Engine.

## Motivation and Context
In a distributed setup File Result Storage Engine can not be used because it saves the file locally. Using Redis for storing the Export Results in such cases.

## How Has This Been Tested?
New Test Cases.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-async/src/test/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+public class RedisResultStorageEngineTest {
+    private static final String HOST = "localhost";
+    private static final int PORT = 6379;
+    private static final int EXPIRATION_SECONDS = 120;
+    private static final boolean EXTENSION_SUPPORT = false;
+    private JedisPool jedisPool;
+    private RedisServer redisServer;
+
+    @BeforeEach
+    public void setup() throws IOException {
```
Comment (aklish) [inline elide-async/src/test/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngineTest.java:34]: We need tests to test the batch reading.

File: elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * Implementation of ResultStorageEngine that stores results on Redis Cluster.
+ * It supports Async Module to store results with Table Export query.
+ */
+@Singleton
+@Slf4j
+@Getter
+public class RedisResultStorageEngine implements ResultStorageEngine {
+    private static final int BATCH_SIZE = 500;
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java:31]: Can this be configurable?

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                        }
+
+                        Iterator<String> itr = jedis.lrange(tableExportID, recordRead[0], end).iterator();
+
+                        // Combine the list into a single string.
+                        while (itr.hasNext()) {
+                            String str = itr.next();
+                            record.append(str).append(System.lineSeparator());
+                        }
+                        recordRead[0] = recordRead[0] + BATCH_SIZE;
```
[the remark this answers is not in the document]
Comment (moizarafat) [reply elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java:109]: Don't think that will work. I have updated the logic with better comments, hopefully that will clarify the logic.

for fewer records than batch size, logic is L-98 to L-100

File: elide-async/src/test/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngineTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        // 3 records > batchSize i.e 2
+        String validOutput = "hi\nhello\nbye";
+        String[] input = validOutput.split("\n");
+
+        storeResults(queryId, Observable.fromArray(input));
+
+        // 2 onnext calls will be returned.
+        // 1st call will have 2 records combined together as one. hi and hello.
+        // 2nd call will have 1 record only. bye.
+        verifyResults("store_results_batch_success", Arrays.asList("hi\nhello", "bye"));
```
Comment (aklish) [inline elide-async/src/test/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngineTest.java:104]: Should there be a newline after hello in this case?

**Review by aklish — DISMISSED** [review]

File: elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java [hunk 5]
```diff
@@ -29,7 +28,7 @@
 @Slf4j
 @Getter
 public class RedisResultStorageEngine implements ResultStorageEngine {
-    @Setter private JedisPool jedisPool;
+    @Setter private UnifiedJedis jedis;
```
Comment (moizarafat) [inline elide-async/src/main/java/com/yahoo/elide/async/service/storageengine/RedisResultStorageEngine.java:31]: Switching to UnifiedJedis so we can switch between JedisCluster and JedisPooled according the environment.

https://github.com/redis/jedis/blob/master/README.md
https://github.com/redis/jedis/releases/tag/v4.0.0

**Review by aklish — APPROVED** [review]

**Pull request merged**